File: src/main/java/org/hisp/dhis/rules/variables/Variable.java
package org.hisp.dhis.rules.variables;

import org.hisp.dhis.antlr.ParserExceptionWithoutContext;
import org.hisp.dhis.rules.RuleExpression;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;

import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;

public class Variable
    extends ScalarFunctionToEvaluate
{
    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        RuleVariableValue variableValue = visitor.getValueMap().get( RuleExpression.getProgramRuleVariable( ctx ) );

        if ( variableValue == null )
        {
            throw new ParserExceptionWithoutContext(
                "Variable " + RuleExpression.getProgramRuleVariable( ctx ) + " not present" );
        }

        return variableValue.value() == null ?
                variableValue.type().defaultValue() : getValueCastedByType( variableValue );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        String variable = RuleExpression.getProgramRuleVariable( ctx );

        if ( visitor.getItemStore().containsKey( variable ) )
        {
            visitor.getItemDescriptions().put( ctx.getText(), visitor.getItemStore().get( variable ).getDisplayName() );

            return visitor.getItemStore().get( RuleExpression.getProgramRuleVariable( ctx ) ).getValueType().getValue();
        }

        throw new ParserExceptionWithoutContext(
            "Variable " + RuleExpression.getProgramRuleVariable( ctx ) + " does not exist" );
    }
}

File: src/jvmMain/kotlin/org/dhis2/ruleengine/variables/Variable.kt
package org.dhis2.ruleengine.variables

import org.dhis2.ruleengine.RuleExpression.Companion.getProgramRuleVariable
import org.dhis2.ruleengine.parser.expression.CommonExpressionVisitor
import org.dhis2.ruleengine.parser.expression.function.ScalarFunctionToEvaluate
import org.hisp.dhis.antlr.ParserExceptionWithoutContext
import org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext

class Variable : ScalarFunctionToEvaluate() {
    override fun evaluate(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variableValue =
            visitor.getValueMap()[getProgramRuleVariable(ctx)]
                ?: throw ParserExceptionWithoutContext(
                    "Variable " + getProgramRuleVariable(ctx) + " not present"
                )
        return variableValue.value ?: variableValue.ruleValueType.defaultValue
    }

    override fun getDescription(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variable = getProgramRuleVariable(ctx)
        if (visitor.getItemStore().containsKey(variable)) {
            visitor.itemDescriptions[ctx.text] = visitor.getItemStore()[variable]!!.displayName
            return visitor.getItemStore()[getProgramRuleVariable(ctx)]!!.valueType.value
        }
        throw ParserExceptionWithoutContext(
            "Variable " + getProgramRuleVariable(ctx) + " does not exist"
        )
    }
}